src/main/java/com/google/sps/servlets/SentimentServlet.java

```diff
@@ -19,71 +22,79 @@
   private Sentiment sentiment;
-  private final String INVALID_INPUT_ERROR = "Request to /sentiment must be an array of Strings.";
-  private final String CONVERT_STRING_ERROR = "Converting request to String failed.";
-  private final String EMPTY_STRING_ERROR = "No comments to analyze. Request must be a non-empty array.";
+  private VideoAnalysis videoAnalysis;
+  private final String INVALID_INPUT_ERROR = "Video is private or does not exist.";
+  private final String COMMENTS_FAILED_ERROR = "Comments could not be retrieved.";
   private final String NLP_API_ERROR = "Language service client failed.";
+  private final String NO_DATA_ERROR = "No comments or captions available to analyze.";
+  private final String READER_ERROR = "Reading video ID failed.";
 
-  public SentimentServlet() throws IOException {
+  public SentimentServlet() throws IOException, GeneralSecurityException {
     languageService = LanguageServiceClient.create();
+    captionService = new Caption();
+    commentService = new CommentService();
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String comments = "";
-    Float commentsScore = null;
+    String json = new Gson().toJson(videoAnalysis);
 
+    response.setContentType("application/json;");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String videoId;
     try {
-      comments = requestToString(request);
-    } catch (JsonSyntaxException e) {
-      System.err.println(e.getMessage());
-      response.sendError(HttpServletResponse.SC_BAD_REQUEST, INVALID_INPUT_ERROR);
-      return;
+      videoId = request.getReader().readLine();
     } catch (IOException e) {
+      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, READER_ERROR);
+      return;
+    }
+    
+    List<String> commentsList = new ArrayList<>();
+    try {
+      commentsList = commentService.getCommentsFromId(videoId);
+    } catch (IllegalArgumentException e) { // video not found
       System.err.println(e.getMessage());
-      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, CONVERT_STRING_ERROR);
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, INVALID_INPUT_ERROR);
       return;
+    } catch (GoogleJsonResponseException e) { // some other problem with requesting comment data
+      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, COMMENTS_FAILED_ERROR);
     }
+    String comments = String.join(". ", commentsList);
+
+    String captions = captionService.getCaptionFromId(videoId);
 
-    if (comments.isEmpty()) {
-      response.sendError(HttpServletResponse.SC_BAD_REQUEST, EMPTY_STRING_ERROR);
+    if (comments.isEmpty() && captions.isEmpty()) {
+      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, NO_DATA_ERROR);
       return;
     }
```

Comment:
Instead of any error for no data, why not just tell the client that as part of the response (VideoAnalysis)? Maybe add another field to VideoAnalysis that conveys this information.